Update docs for LSD granular usage and granular trace data retention filter (#4248)
langchain-ai/docs: src/langsmith/granular-usage.mdx

---
title: Granular billable usage
sidebarTitle: Granular usage
description: Retrieve detailed trace and LangSmith Deployment usage data broken down by workspace, project, user, or API key.
---

<Note>
**Trace usage:** For LangSmith Cloud, granular billable trace data collection started on January 5, 2026. Data is not available for traces ingested before this date.

For self-hosted instances, trace data collection begins when the feature is enabled via the following environment variables, or after [upgrading to a version with it enabled by default](/langsmith/self-hosted-changelog#langsmith-0-13-12).

```env
DEFAULT_ORG_FEATURE_ENABLE_GRANULAR_USAGE_REPORTING=true
GRANULAR_USAGE_TABLE_ENABLED=true
```

> **LangSmith Deployment usage** uses a separate data source — see the Deployment section below.

Reviewer: ```suggestion
**LangSmith Deployment usage** uses a separate data source. For more details, refer to the [LangSmith Deployment section](/langsmith/granular-usage#langsmith-deployment-usage-kind%3Dlangsmith_deployments).
```

</Note>

LangSmith provides granular billable usage APIs that let you retrieve detailed usage data broken down by workspace, project, user, or API key. Two billable domains are supported by the same endpoint, selected via a `kind` query parameter:

- **Trace usage** (`kind=traces`, default) — number of traces ingested.

> - **LangSmith Deployment usage** (`kind=langsmith_deployments`) — nodes executed, agent runs, and agent uptime for [LangSmith Deployment](/langsmith/billing) (formerly LangGraph Platform).

Reviewer: ```suggestion
- **Trace usage** (`kind=traces`, default): number of traces ingested.
- **LangSmith Deployment usage** (`kind=langsmith_deployments`): nodes executed, agent runs, and agent uptime for [LangSmith Deployment](/langsmith/billing).
```

I deleted the LGP ref, because I think we're far enough out from the rebrand of that. But, feel free to keep if you disagree

1. Navigate to **Settings** > **Billing and Usage**
2. Select the **Granular Usage** tab

> 3. Switch between the **LangSmith Traces** and **LangSmith Deployments** sub-tabs to view each domain. The active sub-tab is reflected in the URL (`?tab=traces` or `?tab=deployments`) so the page can be bookmarked and reloads land on the same view.

Reviewer: ```suggestion
3. Switch between the **LangSmith Traces** and **LangSmith Deployments** sub-tabs to view each domain. The active sub-tab is reflected in the URL (`?tab=traces` or `?tab=deployments`) so you can bookmark the page to land on the same view.
```

4. Use the controls to:
   - Select a time range (Last 7 days, 30 days, 3 months, 6 months, 1 year, or custom)
   - Group by workspace, project, user, or API key
   - Filter to specific workspaces
   - On the **LangSmith Traces** tab, optionally filter by retention tier (`All Retention` / `Long-lived only` / `Short-lived only`)
5. Click **Export CSV** to download the data for the active tab.

> Time range and workspace filters are shared across both sub-tabs — switching tabs preserves what you've selected. The **LangSmith Deployments** tab shows three stat cards (Total Nodes Executed / Total Agent Runs / Total Agent Uptime (seconds)) and one chart per metric stacked vertically, since the three metrics use different units.

Reviewer: ```suggestion
Time range and workspace filters are shared across both sub-tabs, switching tabs preserves what you've selected. The **LangSmith Deployments** tab shows three stat cards (Total Nodes Executed / Total Agent Runs / Total Agent Uptime (seconds)) and one chart per metric stacked vertically, since the three metrics use different units.
```

## Query parameters

The granular usage endpoint accepts the following query parameters:

| Parameter | Type | Required | Description |
|-----------|------|----------|-------------|
| `start_time` | datetime | Yes | Start of the time range (ISO 8601 format). |
| `end_time` | datetime | Yes | End of the time range. Must be after `start_time`. |
| `workspace_ids` | array of UUIDs | Yes | Filter results to specific workspaces. |
| `kind` | string | No | `traces` (default) or `langsmith_deployments`. Selects the billable domain. |
| `group_by` | string | No | Dimension to group by. One of: `workspace`, `project`, `user`, `api_key`. Default: `workspace`. |
| `trace_tier` | string | No | Trace-only retention filter: `longlived` or `shortlived`. Omit for all retention. Ignored when `kind=langsmith_deployments`. |

### Day-granular contract

Usage data is aggregated at day granularity. The endpoint normalizes the window to whole days at the API layer:

- `start_time` is floored to its day's UTC midnight.

> - `end_time` is ceiled to the next UTC midnight (no-op when already at midnight).

Reviewer: ```suggestion
- `start_time` is rounded down to its day's UTC midnight.
- `end_time` is rounded up to the next UTC midnight (no-op when already at midnight).
```

- Any day overlapping the requested window is included in full.

A 24-hour window from `2026-01-01T12:00:00Z` to `2026-01-02T12:00:00Z` therefore returns usage for the full Jan 1 and Jan 2 buckets.

### Stride

> The `stride` field in each response indicates the time bucket size used for aggregation, calculated from the requested time range. Daily is the minimum — sub-day windows still bucket at one day.

Reviewer: ```suggestion
The `stride` field in each response indicates the time bucket size used for aggregation, calculated from the requested time range. Daily is the minimum. Sub-day windows still bucket at one day.
```

| Time range | Aggregation | Stride |
|------------|-------------|--------|
| Up to 31 days | Daily | `days: 1` |
| 32–93 days (~3 months) | Weekly | `days: 7` |
| 94–366 days (~1 year) | Monthly | `days: 30` |
| More than 366 days | Yearly | `days: 365` |

### Compatibility

> `kind=langsmith_deployments` combined with `group_by=trace_tier` returns `400 Bad Request` — retention tiers only apply to traces.

Reviewer: ```suggestion
`kind=langsmith_deployments` combined with `group_by=trace_tier` returns `400 Bad Request`. Retention tiers only apply to traces.
```
